src/main/java/dev/morling/demos/logicaldecoding/Main.java

```diff
@@ -51,31 +57,26 @@ public static void main(String[] args) throws Exception {
                 .deserializer(new JsonDebeziumDeserializationSchema()) // converts SourceRecord to JSON String
                 .build();
 
-        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        // for local development, also start the Flink UI on http://localhost:8081
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(new Configuration());
+        env.setParallelism(2);
+        env.getCheckpointConfig().setCheckpointInterval(Time.seconds(5).toMilliseconds());
 
-        env.addSource(sourceFunction)
-                .keyBy(v -> 1L)
-                .map(new AuditMetadataEnrichmentFunction())
-                .print()
-                .setParallelism(1);
+        env.addSource(sourceFunction).forceNonParallel()
+                .map(new AuditMetadataEnrichmentFunction()).forceNonParallel()
+                .print();
 
         env.execute();
     }
 
-    public static class AuditMetadataEnrichmentFunction extends RichMapFunction<String, String> {
+    public static class AuditMetadataEnrichmentFunction extends RichMapFunction<String, String> implements CheckpointedFunction {
 
         private static final long serialVersionUID = 1L;
 
-        private final ObjectMapper mapper;
+        private transient ObjectMapper mapper;
 
-        private transient ValueState<AuditState> auditState;
-
-        public AuditMetadataEnrichmentFunction() {
-            mapper = new ObjectMapper();
-            SimpleModule module = new SimpleModule();
-            module.addDeserializer(Message.class, new MessageDeserializer());
-            mapper.registerModule(module);
-        }
+        private transient ListState<AuditState> auditState;
+        private transient AuditState localAuditState;
```

Comment:
localAuditState is what we are working with regularly, auditState is only used for checkpointing (e.g. making Flink aware of the state)